chain/chain/src/chain.rs

```diff
@@ -2880,6 +2880,28 @@ impl<'a> ChainUpdate<'a> {
             return Err(ErrorKind::EpochOutOfBounds.into());
         }
 
+        // A heuristic to prevent block height to jump too fast towards BlockHeight::max and cause
+        // overflow-related problems
+        let block_height = block.header().height();
+        if block_height > head.height + self.epoch_length * 20 {
+            return Err(ErrorKind::InvalidBlockHeight(block_height).into());
+        }
+
+        // Block is an orphan if we do not know about the previous full block.
+        if !is_next && !self.chain_store_update.block_exists(&block.header().prev_hash())? {
+            // Before we add the block to the orphan pool, do some checks:
+            // 1. Block header is signed by the block producer for height.
+            // 2. Chunk headers in block body match block header.
+            // Not checked:
+            // - Block producer could be slashed
+            // - Chunk header signatures could be wrong
+            if !self.runtime_adapter.partial_verify_orphan_header_signature(&block.header())? {
+                return Err(ErrorKind::InvalidSignature.into());
+            }
+            block.check_validity()?;
```

Comment:
shouldn't this be checked already when the block is received

Reply:
We call `Chain::validate_block` at the beginning, but it validates header first and returns with orphan error. Potentially we can make it call `check_validity` first.

Reply:
I see. It seems that we can just do that.